horace_core/sqw/@axes_block/private/calculate_density_.m

```diff
@@ -0,0 +1,91 @@
+function [dens_nodes,densities,base_cell_size] = calculate_density_(obj,in_data)
+% Convert input datasets defined on centre-points of the axes_block grid into
+% the density data, defined on edges of the axes_block grid.
+%
+% Inputs:
+% in_data -- cellarray of input datasets to calculate density
+%             from.
+%             The size and dimensions of the datasets should
+%             be equal to the dimensions of the axes block
+%             returned by data_nbins property, i.e.:
+%             all(size(dataset{i}) == obj.data_nbins;
+%             datasets contain bin values.
+% Returns:
+% dens_nodes
+%          -- 2D [4,nAxesEdgesPoints] array of axes point positions
+%              where the density is defined
+% densities
+%          -- cellarray of density points calculated in the
+%             density points positions.
+%             Number of cells in the output array is equal to
+%             the number of input datasets
+
+
+
+% build data grid
+[data_nodes,~,npoints_in_base,base_cell_size] = ...
+    obj.get_bin_nodes('-data_to_density');
+gridCX = reshape(data_nodes(1,:),npoints_in_base);
+gridCY = reshape(data_nodes(2,:),npoints_in_base);
+gridCZ = reshape(data_nodes(3,:),npoints_in_base);
+gridCE = reshape(data_nodes(4,:),npoints_in_base);
+%
+
+% build density grid
+base_cell_volume = prod(base_cell_size);
+
+base_cell_size(obj.pax) = base_cell_size(obj.pax)./2;
+[dens_nodes,~,n_ref_points] = obj.get_bin_nodes(base_cell_size);
+
+% provide the coefficient for the future integration over interpolated grid
+% in the form int(grid,a,b) = 0.5*grid_step*sum(signal_i,for a<= i <b);
+densities = cell(numel(in_data),1);
+
+for i = 1:numel(in_data)
+    ref_ds  = convert_data_to_density(obj,in_data{i},base_cell_volume,npoints_in_base);
+    densities{i} = interpn(gridCX,gridCY,gridCZ,gridCE,ref_ds,...
+        dens_nodes(1,:),dens_nodes(2,:),dens_nodes(3,:),dens_nodes(4,:),'linear');
+
+    edges = isnan(densities{i});
+    % for integrated datasets the method generates additional points
+    % contributing to integral. How to account for them?
+    if any(edges(:))
```

Comment:
```suggestion
    if any(edges, 'all')
```

Reply:
> This syntax is valid for MATLAB® versions R2018b and later.
Which is our earliest tested (supported) version.

Reply:
"S = sum (A,'all') computes the sum of all elements of A. This syntax is valid for MATLAB ® versions R2018b and later. example S = sum (A,dim) " -- I still use 2018a on my machine.